### PR TITLE
Single-File: Process bundles in the framework

### LINKED ISSUE #32821: Single-File: Move bundle-processing code into the framework
Implement bundle-processing as described in:
https://github.com/dotnet/designs/blob/master/accepted/2020/single-file/design.md#startup

### PR DESCRIPTION
**Pull request by swaroop-sridhar**
This change implements the host changes proposed in the [design](https://github.com/dotnet/designs/blob/master/accepted/2020/single-file/design.md#startup)

The main changes for single-file bundles are:
* Bundle processing code is moved from apphost to hostpolicy
* HostFxr and HostPolicy process deps.json and runtimeconfig.json files directly from the bundle.
* HostPolicy performs verification wrt deps.json based on the contents of the single-file bundle.
* AppContext.BaseDirectory is set as explained [here](https://github.com/dotnet/designs/blob/master/accepted/2020/single-file/design.md#appcontextbasedirectory)

Currently, all files except deps.json and runtimeconfig.json are extracted to disk.
Once the runtime is able to processing assemblies directly from the bundle, they will no longer be extracted.

Notable details:
* The bundle driver (formarly runner.cpp) is divided into two parts:
   * bundle::info which describes basic information about the bundle available from the headers only
   * bundle::runner which has information about all embedded files, and the ability to drive extraction
  This facilitates linking only parts of the bundle handling code with hostfxr, while all code is linked with hostpolicy.
* The AppHost only links with bundle_marker to identify itself as a single-file bundle.
* If the AppHost is a single-file bundle, it notifies hostfxr using the new hostfxr_main_bundle_startup_info() API
* The HostFxr comminucates the single-file-information with HostPolicy using the host_interface_t structure.

Fixes https://github.com/dotnet/runtime/issues/32821

### CODE REVIEW

File: src/installer/corehost/cli/json/rapidjson/cursorstreamwrapper.h [hunk 1]
```diff
@@ -46,7 +46,7 @@ class CursorStreamWrapper : public GenericStreamWrapper<InputStream, Encoding> {
     // counting line and column number
     Ch Take() {
         Ch ch = this->is_.Take();
-        if(ch == '\n') {
+        if (ch == '\n') {
```
Comment (lpereira) [inline src/installer/corehost/cli/json/rapidjson/cursorstreamwrapper.h:49]: This is a harmless change, but I'd prefer if rapidjson were kept as a pristine copy from upstream, as it's a 3rd-party component.  (If we need to actually fix things in there, I'd prefer that we also sent patches their way instead of maintaining a fork.)

File: src/installer/corehost/cli/json_parser.cpp [hunk 2]
```diff
@@ -130,5 +135,41 @@ bool json_parser_t::parse_stream(pal::istream_t& stream,
...
+#ifdef _WIN32 // _WIN32
+            // Windows doesn't use in-situ parsing. Therefore, pass the memory mapped data directly.
+            // The parser will then copy/convert the data appropriately.
+            parse_json(data);
+            bool result = validate_json(data, location->size, path);
+#else //  _WIN32
+            // On Linux, copy the input data, since m_document.ParseInsitu
+            // requires a mutable data input.
+            realloc_buffer(location->size);
+            memcpy(m_json.data(), data, location->size);
```
Comment (lpereira) [inline src/installer/corehost/cli/json_parser.cpp:163]: I usually see the idiom of using the `swap()` method to initialize a `std::vector` from an array, so I was curious to see how Clang would generate the code. [It's identical](https://gcc.godbolt.org/z/9P7Qn_). I'm impressed. :)

File: src/installer/corehost/cli/bundle/info.cpp [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        return nullptr;
+    }
+
+    // When necessary to map the deps.json or runtimeconfig.json files, we map the whole single-file bundle,
+    // and return the address at the appropriate offset.
+    // This is because:
+    // * The host is the only code that is currently running, and attempting to map the bundle at this time
+    // * Files can only be memory mapped at page-aligned offsets, and in whole page units.
+    //   Therefore, mapping only portions of the bundle will involve align-down/round-up calculations, and associated offset adjustments.
+    //   We choose the simpler approach of rounding to the whole file
```
Comment (lpereira) [inline src/installer/corehost/cli/bundle/info.cpp:102]: Might be a good idea to comment that this might be a problem in 32-bit systems and very large applications due to exhaustion of address space.

File: src/installer/corehost/cli/bundle/info.h [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+                       (the_app->m_deps_json.matches(path) || the_app->m_runtimeconfig_json.matches(path));
+            }
+
+            static const int8_t* map(const pal::string_t& path, const location_t* &location);
+            static void unmap(const int8_t* addr, const location_t* location);
+
+        private:
+            pal::string_t m_path;
+            const location_t *m_location;
+        } json_info;
```
Comment (lpereira) [inline src/installer/corehost/cli/bundle/info.h:53]: `json_info` doesn't seem to be used from a quick glance.  Might want to remove the `config_t()` constructor, then, because method such as `matches()` will reference a garbage pointer

File: src/installer/corehost/cli/bundle/info.h [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+// * Minimal probing and mapping functionality only for the app.runtimeconfig.json and app.deps.json files.
+// bundle::info is used by HostFxr to read the above config files.
+
+namespace bundle
+{
+    struct info_t
+    {
+        struct config_t
+        {
+            config_t() {}
```
Comment (lpereira) [inline src/installer/corehost/cli/bundle/info.h:22]: This doesn't initialize `m_location`, and `config_t::matches()` will dereference that pointer.  When `info_t::m_deps_json`and `info_t::m_runtimeconfig_json` are constructed, they use this constructor here.

File: src/installer/corehost/cli/bundle/manifest.cpp [hunk 6]
```diff
@@ -12,7 +12,9 @@ manifest_t manifest_t::read(reader_t& reader, int32_t num_files)
 
     for (int32_t i = 0; i < num_files; i++)
     {
-        manifest.files.emplace_back(file_entry_t::read(reader));
+        file_entry_t entry = file_entry_t::read(reader);
+        manifest.files.emplace_back(entry);
```
Comment (lpereira) [inline src/installer/corehost/cli/bundle/manifest.cpp:16]: Nit: Is there a benefit for using `emplace_back()`here? Maybe `file_entry_t` should have a move constructor and `std::move`be used instead?

File: src/installer/corehost/cli/bundle/info.cpp [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    pal::string_t runtimeconfig_json_name = get_runtime_config_path(m_base_path, get_filename_without_ext(app_path));
+
+    m_deps_json = config_t(deps_json_name, &m_header.deps_json_location());
+    m_runtimeconfig_json = config_t(runtimeconfig_json_name, &m_header.runtimeconfig_json_location());
+}
+
+const int8_t* info_t::config_t::map(const pal::string_t& path, const location_t* &location)
+{
+    const bundle::info_t* app = bundle::info_t::the_app;
+
```
Comment (lpereira) [inline src/installer/corehost/cli/bundle/info.cpp:82]: Since `the_app` is static and initialized to `nullptr`, might be a good idea to assert if it's not `nullptr` where it's dereferenced

File: src/installer/corehost/cli/bundle/runner.cpp [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    if (entry == nullptr)
+    {
+        return false;
+    }
+
+    bool needs_extraction = entry->needs_extraction();
+    pal::string_t file_base = entry->needs_extraction() ? app->extraction_path() : app->base_path();
+
+    // Reserve space for the path below
+    full_path.reserve(file_base.length() + relative_path.length() + 3);
```
Comment (lpereira) [inline src/installer/corehost/cli/bundle/runner.cpp:73]: What's the `3` for?

File: src/installer/corehost/cli/deps_entry.cpp [hunk 9]
```diff
@@ -6,7 +6,45 @@
...
+
+void deps_entry_t::append_resource_path(pal::string_t& base) const
+{
+    assert(asset_type == asset_types::resources);
+
+    pal::string_t pal_relative_path = asset.relative_path;
+    if (_X('/') != DIR_SEPARATOR)
+    {
+        replace_char(&pal_relative_path, _X('/'), DIR_SEPARATOR);
+    }
```
Comment (lpereira) [inline src/installer/corehost/cli/deps_entry.cpp:32]: `normalize_dir_separator()`?
```suggestion
    pal::string_t pal_relative_path = normalize_dir_separator(asset.relative_path);
```

File: src/installer/corehost/cli/hostmisc/pal.unix.cpp [hunk 10]
```diff
@@ -74,11 +74,16 @@ void* pal::map_file_readonly(const pal::string_t& path, size_t& length)
         close(fd);
         return nullptr;
     }
+    size_t size = buf.st_size;
```
Comment (lpereira) [inline src/installer/corehost/cli/hostmisc/pal.unix.cpp:77]: `size_t` is only for size in memory; for files, you want `off_t`.

File: src/installer/corehost/cli/hostmisc/pal.unix.cpp [hunk 11]
```diff
@@ -74,11 +74,16 @@ void* pal::map_file_readonly(const pal::string_t& path, size_t& length)
...
-    void* address = mmap(nullptr, length, PROT_READ, MAP_SHARED, fd, 0);
+    if (length != nullptr)
+    {
+        *length = size;
+    }
+
+    void* address = mmap(nullptr, size, PROT_READ, MAP_SHARED, fd, 0);
 
-    if(address == nullptr)
+    if (address == nullptr)
```
Comment (lpereira) [inline src/installer/corehost/cli/hostmisc/pal.unix.cpp:86]:
```suggestion
    if (address == MAP_FAILED)
```

File: src/installer/corehost/cli/hostmisc/pal.unix.cpp [hunk 12]
```diff
@@ -74,11 +74,16 @@ void* pal::map_file_readonly(const pal::string_t& path, size_t& length)
...
     {
         trace::error(_X("Failed to map file. mmap(%s) failed with error %d"), path.c_str(), errno);
         close(fd);
```
Comment (lpereira) [inline src/installer/corehost/cli/hostmisc/pal.unix.cpp:89]:
```suggestion
```

File: src/installer/corehost/cli/hostpolicy/hostpolicy_init.cpp [hunk 13]
```diff
@@ -126,6 +127,15 @@ bool hostpolicy_init_t::init(host_interface_t* input, hostpolicy_init_t* init)
         // For the backwards compat case, this will be later initialized with argv[0]
     }
 
+    if (input->version_lo >= offsetof(host_interface_t, app_bundle) + sizeof(input->app_bundle))
+    {
+        if (input->app_bundle != nullptr)
+        {
+            static bundle::runner_t bundle_runner(input->app_bundle);
```
Comment (lpereira) [inline src/installer/corehost/cli/hostpolicy/hostpolicy_init.cpp:134]: I keep reading this as Blade Runner. :)

File: src/installer/corehost/cli/json/rapidjson/cursorstreamwrapper.h (hunk 1 above)
Comment (swaroop-sridhar) [reply src/installer/corehost/cli/json/rapidjson/cursorstreamwrapper.h:49]: OK @lpereira  I'll back off the change from this PR.
I've submitted a PR for rapidjson: https://github.com/Tencent/rapidjson/pull/1685
We should subsequently fix it in our code too. 
I usually do `if(`  `for(` `while(` searches before submitting PRs, and it would be nice to not hit these cases every time.

File: src/installer/corehost/cli/json/rapidjson/document.h [hunk 14]
```diff
@@ -2094,11 +2094,11 @@ class GenericValue {
 
         const SizeType len1 = GetStringLength();
         const SizeType len2 = rhs.GetStringLength();
-        if(len1 != len2) { return false; }
+        if (len1 != len2) { return false; }
```
Comment (swaroop-sridhar) [inline src/installer/corehost/cli/json/rapidjson/document.h:2097]: https://github.com/dotnet/runtime/pull/34274/files/51c0c31d0a9cb26a6e3119bf6609cd300c90c96c#r402006698

File: src/installer/corehost/cli/bundle/runner.cpp (hunk 8 above)
Comment (swaroop-sridhar) [reply src/installer/corehost/cli/bundle/runner.cpp:73]: This is code copied from https://github.com/dotnet/runtime/blob/master/src/installer/corehost/cli/deps_entry.cpp#L32.
Two characters are required for `DIR_SEPARATOR` and null-terminator. Not sure why the third spot is necessary.

File: src/installer/corehost/cli/hostmisc/pal.unix.cpp (hunk 10 above)
Comment (swaroop-sridhar) [reply src/installer/corehost/cli/hostmisc/pal.unix.cpp:77]: I agree that `buf.st_size` is an `off_t` but we're computing `size` to be an input for `mmap()`, which expects a `size_t length` argument .

File: src/installer/corehost/cli/bundle/info.cpp [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+    }
+    else
+    {
+        return nullptr;
+    }
+
+    // When necessary to map the deps.json or runtimeconfig.json files, we map the whole single-file bundle,
+    // and return the address at the appropriate offset.
+    // This is because:
+    // * The host is the only code that is currently running, and attempting to map the bundle at this time
```
Comment (vitek-karas) [inline src/installer/corehost/cli/bundle/info.cpp:100]: This feels like not a finished sentence...

File: src/installer/corehost/cli/bundle/info.cpp [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+    // * Files can only be memory mapped at page-aligned offsets, and in whole page units.
+    //   Therefore, mapping only portions of the bundle will involve align-down/round-up calculations, and associated offset adjustments.
+    //   We choose the simpler approach of rounding to the whole file
+    // * There is no performance limitation due to a larger sized mapping, since we actually only read the pages with relevant contents.
+    // * Files that are too large to be mapped (ex: that exhaust 32-bit virtual address space) are not supported. 
+
+    int8_t* addr = (int8_t*)pal::mmap_read(app->m_bundle_path);
```
Comment (vitek-karas) [inline src/installer/corehost/cli/bundle/info.cpp:107]: I know it's not part of this change yet... do we plan to "cache" the mapping? Once we start loading assemblies from memory I would expect to map the bundle once and hand out pointers from it as required...

File: src/installer/corehost/corehost.cpp [hunk 17]
```diff
@@ -200,23 +187,32 @@ int exe_start(const int argc, const pal::char_t* argv[])
...
     int rc;
-    hostfxr_main_startupinfo_fn main_fn_v2 = reinterpret_cast<hostfxr_main_startupinfo_fn>(pal::get_symbol(fxr, "hostfxr_main_startupinfo"));
-    if (main_fn_v2 != nullptr)
+#if defined(FEATURE_APPHOST)
+    if (bundle_marker_t::is_bundle())
     {
+        hostfxr_main_bundle_startup_info_fn main_fn_v3 = reinterpret_cast<hostfxr_main_bundle_startup_info_fn>(pal::get_symbol(fxr, "hostfxr_main_bundle_startup_info"));
+        if (main_fn_v3 == nullptr)
+        {
+            trace::error(_X("The required library %s does not support single-file apps."), fxr_path.c_str());
```
Comment (vitek-karas) [inline src/installer/corehost/corehost.cpp:196]: What is the experience around this:
* I take a new (5.0) single-file (FDD) app and try to run it on a machine which doesn't have 5.0 installed.
* It fails with this error which doesn't really tell me anything. In addition, for GUI apps the reported `rc` is not recognized as special and there won't be any dialog.

Ideally we want experience similar to hostfxr missing or framework missing. I should get a URL to go to, and for GUI I should get a dialog.
Basically the problem is that this is effectively a version check, which happens before we run FX resolution. So we need it to behave in a very similar manner.

File: src/installer/corehost/corehost.cpp [hunk 18]
```diff
@@ -200,23 +187,32 @@ int exe_start(const int argc, const pal::char_t* argv[])
...
+            rc = StatusCode::CoreHostEntryPointFailure;
+        }
+
         const pal::char_t* host_path_cstr = host_path.c_str();
         const pal::char_t* dotnet_root_cstr = dotnet_root.empty() ? nullptr : dotnet_root.c_str();
         const pal::char_t* app_path_cstr = app_path.empty() ? nullptr : app_path.c_str();
+        int64_t bundle_header_offset = bundle_marker_t::header_offset();
 
-        trace::info(_X("Invoking fx resolver [%s] v2"), fxr_path.c_str());
+        trace::info(_X("Invoking fx resolver [%s] v3"), fxr_path.c_str());
```
Comment (vitek-karas) [inline src/installer/corehost/corehost.cpp:205]: What is the "v3" for? It doesn't really match any version - and the entry points are not versioned in their name. Can we please remove this notion - it's just confusing.
Let's just trace the name of the entry point we're calling. (same below).
Could you also please rename the local variables to not refer to versions...

File: src/installer/corehost/corehost.cpp [hunk 19]
```diff
@@ -200,23 +187,32 @@ int exe_start(const int argc, const pal::char_t* argv[])
...
+        trace::info(_X("Bundle Header Offset: [%lx]"), bundle_header_offset);
 
         hostfxr_set_error_writer_fn set_error_writer_fn = reinterpret_cast<hostfxr_set_error_writer_fn>(pal::get_symbol(fxr, "hostfxr_set_error_writer"));
         {
             propagate_error_writer_t propagate_error_writer_to_hostfxr(set_error_writer_fn);
 
-            rc = main_fn_v2(argc, argv, host_path_cstr, dotnet_root_cstr, app_path_cstr);
+            rc = main_fn_v3(argc, argv, host_path_cstr, dotnet_root_cstr, app_path_cstr,bundle_header_offset);
 
             if (trace::get_error_writer() != nullptr && rc == static_cast<int>(StatusCode::FrameworkMissingFailure) && !set_error_writer_fn)
```
Comment (vitek-karas) [inline src/installer/corehost/corehost.cpp:217]: This is not necessary due to the above "version" check. This was introduced here: https://github.com/dotnet/runtime/pull/1222
The reason was to get a good error message for UI apps when running 3.0 apps on 2.0 only hostfxr which doesn't support error writer redirection. In this case the failure will happen earlier and if we get here it will be with >=5.0 hostfxr which will support the fully featured error reporting.

Maybe add comments describing this reasoning here and to the case below for 3.0 app on 2.0 hostfxr.

**Review by vitek-karas — CHANGES_REQUESTED** [review]
I have yet to look at tests. The most important feedback is about the UX when we run 5.0 single-file on machine which doesn't have 5.0. We need to get that right.

File: src/installer/corehost/cli/bundle/info.cpp (hunk 16 above)
Comment (swaroop-sridhar) [reply src/installer/corehost/cli/bundle/info.cpp:107]: There's no plan to cache the mapping. 
* Closing the mapping enables renaming the app while it is running on Windows -- an idiom used to update apps in-place.
* The PE loader itself maps each section of the bundled assemblies at aligned addresses.

The only item that is cached is the bundle-manifest, so that bundle-probes from the runtime can be answered without need for re-processing the bundle headers.

File: src/installer/corehost/corehost.cpp (hunk 17 above)
Comment (swaroop-sridhar) [reply src/installer/corehost/corehost.cpp:196]: I've fixed this, here's a sample output:
```
winhello.exe
The required library C:\Program Files\dotnet\host\fxr\5.0.0-preview.3.20169.1\hostfxr.dll does not support single-file apps.
  _ To run this application, you need to install a newer version of .NET Core.
  - https://aka.ms/dotnet-core-applaunch?missing_runtime=true&arch=x64&rid=win10-x64
```
image

File: src/installer/corehost/cli/bundle/info.cpp (hunk 16 above)
Comment (vitek-karas) [reply src/installer/corehost/cli/bundle/info.cpp:107]: Good point - I forgot that runtime will create its own mapping anyway.

File: src/installer/corehost/corehost.cpp [hunk 20]
```diff
@@ -84,6 +83,14 @@ bool is_exe_enabled_for_execution(pal::string_t* app_dll)
 #define CURHOST_EXE
 #endif
 
+void need_newer_framework_error()
+{
+    pal::string_t url = get_download_url();
+    trace::error(_X("  _ To run this application, you need to install a newer version of .NET Core."));
+    trace::error(_X(""));
+    trace::error(_X("  - %s"), url.c_str());
```
Comment (vitek-karas) [inline src/installer/corehost/corehost.cpp:91]: Can you please append the apphost version to the URL? The problem is that without it there's no version information in the URL and the web server will have to way of telling at least roughly which version to offer. Once we ship 6.0 that would probably become the default and 5.0 apps would not work with it (default roll forward is Minor only). It would require two installations (6.0 first - which would provide better error reporting as we would load hostfxr and then 5.0 to actually run the app).

See this change which improved this for the other cases: https://github.com/dotnet/runtime/pull/33602

File: src/installer/corehost/corehost.cpp (hunk 17 above)
Comment (vitek-karas) [reply src/installer/corehost/corehost.cpp]: Can you please double-check that the URL which the dialog leads to contains apphost version information?

File: src/installer/test/Assets/TestProjects/AppWithSubDirs/Program.cs [hunk 21]
```diff
@@ -3,6 +3,7 @@
 // See the LICENSE file in the project root for more information.
 
 using System;
+using System.Diagnostics;
```
Comment (vitek-karas) [inline src/installer/test/Assets/TestProjects/AppWithSubDirs/Program.cs:6]: Nit: probably not needed

**Review by vitek-karas — COMMENTED** [review]
I think I 'm struggling to understand what is the hostpolicy/coreclr interface. In the design doc https://github.com/dotnet/designs/blob/master/accepted/2020/single-file/design.md#dependency-resolution it says:
>The assemblies bundled within the single-file are not enumerated in TRUSTED_PLATFORM_ASSEMBLIES. The absolute-path of assemblies on disk are listed in TRUSTED_PLATFORM_ASSEMBLIES as usual.

But if I read the code correctly it will populate TPA with full paths even for bundled assemblies - relative to the app's base.

Similar comment on resource paths.

File: src/installer/corehost/cli/deps_entry.cpp [hunk 22]
```diff
@@ -21,28 +21,29 @@ static pal::string_t normalize_dir_separator(const pal::string_t& path)
...
-bool deps_entry_t::to_path(const pal::string_t& base, bool look_in_base, pal::string_t* str) const
+// -----------------------------------------------------------------------------
+// Given a "base" directory, determine the resolved path this this file.
+//
+// If this file exists within the single-file bundle candidate is:
+//  * If the file was extracted to disk, the full-path to the extracted file.
+//  * Otherwise, the path within the bundle, relative to the "base" directory.
+//    The runtime expects the entries in the TPA,  NativeDllSearchDirectories, etc 
+//    to be absolute paths. Therefore, the relative-paths within the bundle are
+//    expressed as absolute paths with respect to the location of the bundle-file.
```
[the remark this answers is not in the document]
Comment (swaroop-sridhar) [reply src/installer/corehost/cli/deps_entry.cpp:32]: But I agree that this comment is confusing, because it tries to talk about both the current and future state of the machine. I'll fix it to only say that the assembly is expected to be found at the extracted location.

File: src/installer/corehost/corehost.cpp (hunk 17 above)
Comment (swaroop-sridhar) [reply src/installer/corehost/corehost.cpp]: Yes the link generated now is:
https://aka.ms/dotnet-core-applaunch?missing_runtime=true&arch=x64&rid=win10-x64&apphost_version=5.0.0-dev

File: src/installer/corehost/cli/host_interface.h [hunk 23]
```diff
@@ -58,6 +59,7 @@ struct host_interface_t
     const pal::char_t* host_info_host_path;
     const pal::char_t* host_info_dotnet_root;
     const pal::char_t* host_info_app_path;
+    const bundle::info_t *app_bundle;
```
Comment (elinor-fung) [inline src/installer/corehost/cli/host_interface.h:62]: `bundle::info_t` does not look to be plain data, so using it cross-binary like this where hostpolicy reads the class allocated by hostfxr could lead to compat issues.

File: src/installer/corehost/cli/fxr/hostpolicy_resolver.cpp [hunk 24]
```diff
@@ -12,6 +12,7 @@
 #include <utils.h>
 
 #include "json_parser.h"
+#include "bundle/info.h"
```
Comment (elinor-fung) [inline src/installer/corehost/cli/fxr/hostpolicy_resolver.cpp:15]: Not needed?

File: src/installer/corehost/cli/deps_entry.cpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (vitek-karas) [reply src/installer/corehost/cli/deps_entry.cpp]: So if I read it correctly, this is a point-in-time behavior.
Currently it is expected to actually extract all assemblies on disk. And so we calculate the path as relative to the app-base.
In the future, these will not be extracted and thus the code will change to not include them in TPA (and the logic how they're resolved will have to adapt as well). Right?

**Review by vitek-karas — APPROVED** [review]
Looks good - the one pending question I have is the RID specific asset handling.

File: src/installer/corehost/cli/deps_entry.cpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (swaroop-sridhar) [reply src/installer/corehost/cli/deps_entry.cpp]: Yes, this is correct.

File: src/installer/corehost/cli/host_interface.h (hunk 23 above)
Comment (swaroop-sridhar) [reply src/installer/corehost/cli/host_interface.h]: Thanks @elinor-fung . I've fixed this in https://github.com/dotnet/runtime/pull/34274/commits/12b63bae12ab435120b8a9b56165a86d6563406f

File: src/installer/corehost/cli/hostpolicy/deps_resolver.cpp [hunk 25]
```diff
@@ -343,6 +343,12 @@ bool deps_resolver_t::probe_deps_entry(const deps_entry_t& entry, const pal::str
                 else
                 {
                     // Non-rid assets, lookup in the published dir.
```
[the remark this answers is not in the document]
Comment (swaroop-sridhar) [reply src/installer/corehost/cli/hostpolicy/deps_resolver.cpp:345]: @vitek-karas, regarding rid-specific assets: 
* I've added the bundle-check in the rid-specific assets case too, as you've recommended.
* However, when an app is published specific to a RID, my understanding is that the SDK will never create assets for other RIDs. The SDK will find the best compatible RID assets and copy them to the publish directory.  So, there won't be a `<rid>\publish\runtimes\<rid>` structure.

For example, if we have a package (libuv 1.9.0 is a very good example)
```
runtimes
   |- debian-x64/native/lib.so
   |- rhel-x64/native/lib.so
   |- win-x64/native/lib.dll
```
`dotnet publish` will generate the above directory structure within the `bin/.net5/publish` directory.

However, `dotnet publish -r debian-x64 --self-contained=false`  will generate `bin/.net5/debian-x64/publish` that *only* contains `lib.so` from `runtime/debian-x64/native`.

`dotnet publish -r ubuntu-x64 --self-contained=false`  will generate `bin/.net5/ubuntu-x64/publish` that contains `lib.so` from `runtime/debian-x64/native`.

`dotnet publish -r linux-x64 --self-contained=false`  will generate `bin/.net5/linux-x64/publish` without `lib.so`

**Comment by swaroop-sridhar** [comment]
> the one pending question I have is the RID specific asset handling.

I've now answered this in https://github.com/dotnet/runtime/pull/34274#discussion_r405265681

File: src/installer/corehost/cli/hostpolicy/deps_resolver.cpp (hunk 25 above)
[the remark this answers is not in the document]
Comment (vitek-karas) [reply src/installer/corehost/cli/hostpolicy/deps_resolver.cpp:345]: > dotnet publish -r linux-x64 --self-contained=false will generate bin/.net5/linux-x64/publish without lib.so

This sounds like a problem/bug. I assume the app would not run in that case at all... right?

**Review by vitek-karas — APPROVED** [review]
Looks good.

Please followup on the issue with RID-specific assets. I do agree that vast majority of the apps won't have the problem. If SDK will never add a RID-specific asset to a RID-specific build, then this is not an issue for host/runtime (yet), but it seems to be an SDK issue.

File: src/installer/corehost/cli/hostpolicy/deps_resolver.cpp (hunk 25 above)
[the remark this answers is not in the document]
Comment (swaroop-sridhar) [reply src/installer/corehost/cli/hostpolicy/deps_resolver.cpp:345]: I agree that there is an underlying problem, because the app may fail to run because of missing assemblies in this situation. However, I'm not sure it is an SDK bug, but rather a product issue. 

The missing RID-asset (here `linux-x64`) in a package may actually mean that the library (`lib.so`) is unnecessary for that platform -- although it is weird that the lib is not necessary at a higher RID but necessary in a lower RID. This is somewhat similar to missing an assembly that is only loaded at runtime for reflection -- where the SDK cannot know the exact requirements. 

@dsplaisted  @wli3: Should we consider guards in the SDK to check for this case? (ex: missing assets at a parent RID)? Thanks.

[the remark this answers is not in the document]
Comment (dsplaisted) [reply src/installer/corehost/cli/hostpolicy/deps_resolver.cpp:345]: @swaroop-sridhar I have a feeling trying to add guards for this type of thing would be more complex and complicated than would be warranted for the benefit we would get out of it.  But you can file an issue in the SDK repo if you want to discuss it further.

[the remark this answers is not in the document]
Comment (swaroop-sridhar) [reply src/installer/corehost/cli/hostpolicy/deps_resolver.cpp:345]: https://github.com/dotnet/sdk/issues/11206
